admin/services/restart.service.js

```diff
@@ -0,0 +1,39 @@
+const moment = require('moment')
+const pupilDataService = require('../services/data-access/pupil.data.service')
+const schoolDataService = require('../services/data-access/school.data.service')
+const pupilIdentificationFlagService = require('../services/pupil-identification-flag.service')
+const pinValidator = require('../lib/validator/pin-validator')
+
+const restartService = {}
+
+/**
+ * Get pupils who are eligible for restart
+ * @param schoolId
+ * @returns {Array}
+ */
+restartService.getPupils = async (schoolId) => {
+  const school = await schoolDataService.findOne({_id: schoolId})
+  if (!school) throw new Error(`School [${schoolId}] not found`)
+  let pupils = await pupilDataService.getSortedPupils(schoolId, 'lastName', 'asc')
+  pupils = pupils
+    .filter(p => restartService.isEligiblePupil(p))
+    .map(({ _id, pin, dob, foreName, middleNames, lastName }) =>
+      ({ _id, pin, dob: moment(dob).format('DD MMM YYYY'), foreName, middleNames, lastName }))
+  pupils = pupilIdentificationFlagService.addIdentificationFlags(pupils)
+  return pupils
+}
+
+/**
+ * Determine if pupil is eligible for restart
+ * @param p
+ * @returns {Boolean}
+ */
+restartService.isEligiblePupil = (p) => {
+  // This is undergoing changes as the count won't be attached to the pupil collection
+  let restartCount = p.restartCount || 0
+  if (restartCount >= 2 || p.attendanceCode) return false
+  const hasExpiredToday = p.pinExpiresAt && moment(p.pinExpiresAt).isSame(moment.now(), 'day')
```

Comment:
If the pupil pin expires today, and this is called before the pin expiry time, then won't `hasExpiredToday` be true even though the pupil pin is still valid?

Reply:
But this boolean is only there to verify that the of the pinExpiration is today.
The check on the line below will determine based on today the result from the validation of the pin.